### PR TITLE
Implement "microk8s images import" command

### PR DESCRIPTION
**Pull request by neoaggelos**
#### Summary

Implement `microk8s images import` command, which can be used to import OCI images to all nodes of a MicroK8s cluster.

Assumes https://github.com/canonical/microk8s-cluster-agent/pull/4

#### Changes

- Refactor existing code in `distributed_op`: Pull out the code that retrieves the list of all current nodes and their callback tokens
- Implement `do_image_import`, similar to the existing `do_op`.
- Implement `microk8s.images import` command, importing images to the whole cluster.

#### Testing

Manually

#### Possible Regressions

Please check if I have missed something in the original code where the list of nodes is retrieved. There was a lot of duplicate code which I have tried to refactor.

I have tried to preserve the error messages in all cases.

#### Checklist

* [X] Read the [contributions](https://github.com/canonical/microk8s/blob/master/CONTRIBUTING.md) page.
* [X] Submitted the [CLA form](https://ubuntu.com/legal/contributors/agreement), if you are a first time contributor.
* [X] The introduced changes are covered by unit and/or integration tests.

### CODE REVIEW

File: scripts/cluster/distributed_op.py [hunk 1]
```diff
@@ -15,90 +15,123 @@
...
-                        json=remote_op,
-                        verify=False,
-                    )
-                    if res.status_code != 200:
-                        print(
-                            "Failed to perform a {} on node {}".format(
-                                remote_op["action_str"], node_ep
-                            )
-                        )
+                    subprocess.check_call([KUBECTL, "get", "node", host])
```
Comment (ktsakalozos) [inline scripts/cluster/distributed_op.py:65]: We probably need to send stdout and stderr in devnul here or else we get an output like this one:
```
$ sudo microk8s.enable dns
Infer repository core for addon dns
Enabling DNS
Applying manifest
serviceaccount/coredns created
configmap/coredns created
deployment.apps/coredns created
service/kube-dns created
clusterrole.rbac.authorization.k8s.io/coredns created
clusterrolebinding.rbac.authorization.k8s.io/coredns created
Restarting kubelet
Adding argument --cluster-domain to nodes.
NAME   STATUS   ROLES    AGE    VERSION
u2     Ready    <none>   5m1s   v1.24.1-2+d1d054585eedd9
NAME   STATUS   ROLES    AGE   VERSION
u3     Ready    <none>   39s   v1.24.1-2+d1d054585eedd9
Adding argument --cluster-dns to nodes.
NAME   STATUS   ROLES    AGE    VERSION
u2     Ready    <none>   5m2s   v1.24.1-2+d1d054585eedd9
NAME   STATUS   ROLES    AGE   VERSION
u3     Ready    <none>   40s   v1.24.1-2+d1d054585eedd9
Restarting nodes.
NAME   STATUS   ROLES    AGE    VERSION
u2     Ready    <none>   5m5s   v1.24.1-2+d1d054585eedd9
NAME   STATUS   ROLES    AGE   VERSION
u3     Ready    <none>   43s   v1.24.1-2+d1d054585eedd9
DNS is enabled

```

Comment (neoaggelos) [reply scripts/cluster/distributed_op.py]: good catch

**Review by ktsakalozos — CHANGES_REQUESTED** [review]

File: scripts/cluster/distributed_op.py [hunk 2]
```diff
@@ -15,90 +15,127 @@
...
+        stdout = subprocess.check_output([KUBECTL, "get", "node", "-o", "json"])
+        info = json.loads(stdout)
+        for node_info in info["items"]:
+            node_ip = get_internal_ip_from_get_node(node_info)
+            if not include_self and is_same_server(hostname, node_ip):
+                continue
+
+            nodes.append(("{}:25000".format(node_ip), token.rstrip()))
     else:
         with open(callback_tokens_file, "r+") as fp:
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply scripts/cluster/distributed_op.py:59]: Do we have a callback token in this case on the node where this command is running? Could we add 127.0.0.1:25000 manually?

File: scripts/cluster/distributed_op.py [hunk 3]
```diff
@@ -15,90 +15,127 @@
...
+            res = requests.post(
+                "https://{}/{}/image/import".format(node_ep, CLUSTER_API_V2),
+                data=image_data,
+                headers={
+                    "x-microk8s-callback-token": token,
+                },
+                verify=False,
+            )
+
+            if res.status_code != 200:
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply scripts/cluster/distributed_op.py:134]: this is by design, in order to proceed with the rest of the cluster nodes, even if some are (temporarily?) unavailable

File: scripts/cluster/distributed_op.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ktsakalozos) [reply scripts/cluster/distributed_op.py:65]: I do not know what would happen if we did that.

Overall you have already seen that the distributed ops is doing the operation to all the other nodes not the current one. I have seen in the past operations that are applied differently on the local node and on the remote ones.

[the remark this answers is not in the document]
Comment (neoaggelos) [reply scripts/cluster/distributed_op.py:65]: I have updated the code to retrieve the local callback token (creating it, if need be) and include the local cluster agent as well. Is there an alternative suggestion?

[the remark this answers is not in the document]
Comment (ktsakalozos) [reply scripts/cluster/distributed_op.py:65]: Here is a new error. This time on a single node cluster.
```
ubuntu@u1:~$ sudo microk8s.disable ha-cluster --force
Infer repository core for addon ha-cluster
Reverting to a non-HA setup
Generating new cluster certificates.
Waiting for node to start. .  
Enabling flanneld and etcd
HA disabled
ubuntu@u1:~$ sudo microk8s.images import /var/tmp/ubuntu.tar 
Traceback (most recent call last):
  File "/snap/microk8s/x1/scripts/cluster/images.py", line 29, in <module>
    images(prog_name="microk8s images")
  File "/snap/microk8s/x1/usr/lib/python3/dist-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/snap/microk8s/x1/usr/lib/python3/dist-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/snap/microk8s/x1/usr/lib/python3/dist-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/snap/microk8s/x1/usr/lib/python3/dist-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/snap/microk8s/x1/usr/lib/python3/dist-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/snap/microk8s/x1/scripts/cluster/images.py", line 25, in import_images
    do_image_import(image_data)
  File "/snap/microk8s/x1/scripts/cluster/distributed_op.py", line 123, in do_image_import
    endpoints = get_cluster_agent_endpoints(include_self=True)
  File "/snap/microk8s/x1/scripts/cluster/distributed_op.py", line 65, in get_cluster_agent_endpoints
    with open(callback_tokens_file, "r+") as fp:
FileNotFoundError: [Errno 2] No such file or directory: '/var/snap/microk8s/x1/credentials/callback-tokens.txt'
```

**Pull request merged**